Add Saml2AuthenticationRequestRepository

marcusdacoregio opened the pull request:
Closes gh-9185

Related #9174

<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->

<!--
Before creating new features, we recommend creating an issue to discuss the feature. This ensures that everyone is on the same page before extensive work is done.

Thanks for contributing to Spring Security. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with gh-).
-->

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2AuthenticationToken.java:
```diff
@@ -38,6 +38,32 @@
+	 */
+	public Saml2AuthenticationToken(RelyingPartyRegistration relyingPartyRegistration, String saml2Response,
+			AbstractSaml2AuthenticationRequest authenticationRequest) {
+		super(Collections.emptyList());
```
Ideally, this constructor would call the other constructor to reduce maintenance.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/HttpSessionSaml2AuthenticationRequestRepository.java:
```diff
@@ -0,0 +1,72 @@
+	@Override
+	public void saveAuthenticationRequest(AbstractSaml2AuthenticationRequest authenticationRequest,
+			HttpServletRequest request, HttpServletResponse response) {
+		HttpSession httpSession = request.getSession(false);
```
If someone is proactively saving it, I think it's reasonable to call `request.getSession()` instead.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/HttpSessionSaml2AuthenticationRequestRepository.java:
```diff
@@ -0,0 +1,72 @@
+	public void saveAuthenticationRequest(AbstractSaml2AuthenticationRequest authenticationRequest,
+			HttpServletRequest request, HttpServletResponse response) {
+		HttpSession httpSession = request.getSession(false);
+		if (httpSession == null || authenticationRequest == null) {
```
I think if the request is null, that should be the equivalent of removing the request from the session. This aligns with the Servlet spec, which says about `setAttribute`:

```javadoc
If the value passed in is null, this has the same effect as calling removeAttribute().
```

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/HttpSessionSaml2AuthenticationRequestRepository.java:
```diff
@@ -0,0 +1,72 @@
+	@Override
+	public AbstractSaml2AuthenticationRequest removeAuthenticationRequest(HttpServletRequest request,
+			HttpServletResponse response) {
+		AbstractSaml2AuthenticationRequest saml2AuthenticationRequest = this.loadAuthenticationRequest(request);
```
Some Spring Security classes still add `this.` to method calls, but it is not part of the style guide. Will you please remove `this.` from method calls?

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/HttpSessionSaml2AuthenticationRequestRepository.java:
```diff
@@ -0,0 +1,72 @@
+			return null;
+		}
+		HttpSession httpSession = request.getSession(false);
+		httpSession.removeAttribute(this.saml2AuthnRequestAttributeName);
```
This might throw an NPE since `getSession(boolean)` may return `null`. I'd recommend `request.getSession()` instead since if `loadAuthenticationRequest` returns non-null, then the session is present.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java:
```diff
@@ -95,6 +101,7 @@ protected boolean requiresAuthentication(HttpServletRequest request, HttpServlet
 	public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response)
 			throws AuthenticationException {
 		Authentication authentication = this.authenticationConverter.convert(request);
+		this.authenticationRequestRepository.removeAuthenticationRequest(request, response);
```
It seems that if `authentication == null` this might be an invalid request. Should the authnrequest be removed after it's clear that `authentication != null`?

marcusdacoregio replied:
Yes, it should.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java:
```diff
@@ -103,4 +110,16 @@ public Authentication attemptAuthentication(HttpServletRequest request, HttpServ
+	 */
+	public void setAuthenticationRequestRepository(
+			Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> authenticationRequestRepository) {
+		this.authenticationRequestRepository = authenticationRequestRepository;
```
Please check for null. `Assert.notNull` is handy here.

jzheaux replied:
This method also probably needs to pass the provided request repository down to the authentication converter, if the converter is of the correct type.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/Saml2AuthenticationTokenConverter.java:
```diff
@@ -62,6 +66,23 @@ public Saml2AuthenticationTokenConverter(
+	 */
+	public Saml2AuthenticationTokenConverter(
+			Converter<HttpServletRequest, RelyingPartyRegistration> relyingPartyRegistrationResolver,
+			Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> authenticationRequestRepository) {
```
Usually, it's preferred to either have a setter or a constructor parameter, but not both. In this case, since `Saml2AuthenticationRequestRepository` is optional, I'd recommend removing the constructor parameter.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/Saml2AuthenticationTokenConverter.java:
```diff
@@ -50,6 +52,8 @@
 
 	private final Converter<HttpServletRequest, RelyingPartyRegistration> relyingPartyRegistrationResolver;
 
+	private Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> authenticationRequestRepository;
```
Is there a reason that this should not be set initially to `HttpSessionAuthenticationRequestRepository`?

marcusdacoregio replied:
There isn't, I'll initialize it with `HttpSessionAuthenticationRequestRepository`.

jzheaux commented on saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/servlet/HttpSessionSaml2AuthenticationRequestRepositoryTest.java:
```diff
@@ -0,0 +1,133 @@
+/**
+ * @author Marcus Da Coregio
+ */
+public class HttpSessionSaml2AuthenticationRequestRepositoryTest {
```
Please call this `Tests` instead of `Test`.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2AuthenticationToken.java:
```diff
@@ -38,6 +38,32 @@
 
 	private final String saml2Response;
 
+	private AbstractSaml2AuthenticationRequest authenticationRequest;
```
Since this won't change, I think it can be `final`.

marcusdacoregio replied:
If I do this in the new constructor:
```java
public Saml2AuthenticationToken(RelyingPartyRegistration relyingPartyRegistration, String saml2Response,
			AbstractSaml2AuthenticationRequest authenticationRequest) {
	this(relyingPartyRegistration, saml2Response);
	this.authenticationRequest = authenticationRequest;
}
```
I can't make it `final` since it could be initialized inside the first constructor.

jzheaux replied:
I think the constructors should be inverted. For example, do:

```java
public Saml2AuthenticationToken(RelyingPartyRegistration relyingPartyRegistration, String saml2Response) {
    this(relyingPartyRegistration, saml2Response, null);
```

marcusdacoregio replied:
Nice!

jzheaux reviewed (changes requested):
Thanks for the PR, @marcusdacoregio!

In addition to the comments, please also add DSL support and docs. At a minimum, the DSL support should look for a bean of type `Saml2AuthorizationRequestRepository` and wire it in accordingly.

marcusdacoregio commented:
Thanks for your review @jzheaux. I've updated the PR with your suggestions and the DSL and docs addition.

jzheaux commented on docs/manual/src/docs/asciidoc/_includes/servlet/saml2/saml2-login.adoc:
```diff
@@ -1610,3 +1610,33 @@ http {
+
+The default implementation of `Saml2AuthenticationRequestRepository` is `HttpSessionSaml2AuthenticationRequestRepository`, which stores the `AbstractSaml2AuthenticationRequest` in the `HttpSession`.
+
+If you have a custom implementation of `Saml2AuthenticationRequestRepository`, you may configure it by exposing it as a *Bean* as shown in the following example:
```
Instead of *emphasizing* words like "Bean", we should mark them as code, like `@Bean`.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java:
```diff
@@ -100,7 +106,34 @@ public Authentication attemptAuthentication(HttpServletRequest request, HttpServ
+
+	private void setAuthenticationRequestRepositoryIntoAuthenticationConverter(
+			Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> authenticationRequestRepository) {
+		if (!this.authenticationConverter.getClass().isAssignableFrom(Saml2AuthenticationTokenConverter.class)) {
```
Could you instead do:

```java
if (this.authenticationConverter instanceof Saml2AuthenticationTokenConverter) {
```

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java:
```diff
@@ -83,6 +88,7 @@ public Saml2WebSsoAuthenticationFilter(AuthenticationConverter authenticationCon
 		this.authenticationConverter = authenticationConverter;
 		setAllowSessionCreation(true);
 		setSessionAuthenticationStrategy(new ChangeSessionIdAuthenticationStrategy());
+		setAuthenticationRequestRepository(new HttpSessionSaml2AuthenticationRequestRepository());
```
I realize that this is what was done with `setSessionAuthenticationStrategy`; however, it's preferable to not call overrideable methods in a constructor. It may be better to do:

```java
private Saml2AuthenticationRequestRepository authenticationRequestRepository =
    new HttpSessionAuthenticationRequestRepository();
```

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/Saml2AuthenticationTokenConverter.java:
```diff
@@ -74,7 +79,27 @@ public Saml2AuthenticationToken convert(HttpServletRequest request) {
+	 */
+	public void setAuthenticationRequestRepository(
+			Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> authenticationRequestRepository) {
+		this.authenticationRequestRepository = authenticationRequestRepository;
```
Let's please check for null here, using `Assert.notNull`.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/Saml2AuthenticationTokenConverter.java:
```diff
@@ -74,7 +79,27 @@ public Saml2AuthenticationToken convert(HttpServletRequest request) {
+	}
+
+	private AbstractSaml2AuthenticationRequest loadAuthenticationRequest(HttpServletRequest request) {
+		if (this.authenticationRequestRepository == null) {
```
Instead of this null check, you might consider following the Null Object pattern, like so:

```java
private Function<HttpServletRequest, AbstractSaml2AuthenticationRequest> loader = (request) -> null;

// ...

public void setAuthenticationRequestRepository(Saml2AuthenticationRequestRepository requests) {
    // ...
    this.loader = requests::loadAuthenticationRequest;
}
```

This improves readability by removing identation and conditionals.

jzheaux reviewed (changes requested):
Thanks for the updates, @marcusdacoregio! I've left some additional feedback about the changes you submitted.

marcusdacoregio commented on config/src/main/java/org/springframework/security/config/annotation/web/configurers/saml2/Saml2LoginConfigurer.java:
```diff
@@ -339,8 +351,19 @@ private AuthenticationRequestEndpointConfig() {
+
+		private Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> getAuthenticationRequestRepository(
+				B http) {
+			return getSharedOrBean(http, Saml2AuthenticationRequestRepository.class);
```
Actually, there isn't. I was just considering that the Filter already initializes an `HttpSessionSaml2AuthenticationRequestRepository` when constructed

marcusdacoregio commented:
Thanks for your effort on the review @jzheaux. I've updated the PR with your suggestions.

jzheaux commented on saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/Saml2AuthenticationTokenConverter.java:
```diff
@@ -50,6 +54,10 @@
 
 	private final Converter<HttpServletRequest, RelyingPartyRegistration> relyingPartyRegistrationResolver;
 
+	private Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> authenticationRequestRepository = new HttpSessionSaml2AuthenticationRequestRepository();
```
It would be better to remove this reference since it would otherwise need to stay in sync with the `loader` member.

For example, in your constructor you could do:

```java
this.loader = new HttpSessionSaml2AuthenticationRequestFactory()::loadAuthenticationRequest;
```

jzheaux commented on config/src/main/java/org/springframework/security/config/annotation/web/configurers/saml2/Saml2LoginConfigurer.java:
```diff
@@ -33,13 +33,16 @@
 import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
```
Please update this copyright and any others that are not yet updated to 2021.

jzheaux reviewed (changes requested):
Thanks, @marcusdacoregio! I've left some additional feedback inline.

marcusdacoregio commented:
Thanks @jzheaux, I've updated the PR based on your feedback.

jzheaux commented on docs/manual/src/docs/asciidoc/_includes/servlet/saml2/saml2-login.adoc:
```diff
@@ -1610,3 +1610,33 @@ http {
+[[servlet-saml2login-store-authn-request]]
+=== Storing the Authentication Request
+
+The `Saml2AuthenticationRequestRepository` is responsible for the persistence of the `AbstractSaml2AuthenticationRequest` from the time the Authentication Request is initiated to the time the Authentication Response is received.
```
You might consider linking back to `[[servlet-saml2login-sp-initiated-factory]]` and `[[servlet-saml2login-authenticate-responses]]` to help folks back to the sections that give an overview of how Spring Security handles AuthnRequests and SAML Responses.

jzheaux reviewed (changes requested):
Things are really taking shape now, @marcusdacoregio, nice work.

I've left some remaining feedback about JavaDoc, docs, and tests.

marcusdacoregio commented on saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilterTests.java:
```diff
@@ -84,4 +93,65 @@ public void attemptAuthenticationWhenRegistrationIdDoesNotExistThenThrowsExcepti
+	@Test
+	public void setAuthenticationRequestRepositoryWhenExpectedAuthenticationConverterTypeThenSetLoaderIntoConverter() {
+		Saml2AuthenticationTokenConverter authenticationConverterMock = mock(Saml2AuthenticationTokenConverter.class);
+		MockSaml2AuthenticationRequestRepository mockAuthenticationRequestRepository = new MockSaml2AuthenticationRequestRepository();
```
Actually, there isn't. I went for the dedicated mock but turns out that it's the same as a Mockito mock 😃

marcusdacoregio commented:
Thank you again for the review @jzheaux. I've updated the PR.

About the docs, I've updated it to use the SAML terms, like `AuthnRequest` and `SAMLResponse`, aligned with the rest of the docs as you mentioned.

jzheaux commented:
Thanks, @marcusdacoregio! This is now merged into `main`.
